esp/esdllib/esdl_script.cpp

```diff
@@ -93,10 +93,163 @@ class OptionalCriticalBlock
-void createEsdlTransformChooseOperations(IArrayOf<IEsdlTransformOperation> &operations, IXmlPullParser &xpp, const StringBuffer &prefix, bool withVariables, bool ignoreCodingErrors, IEsdlFunctionRegister *functionRegister);
-typedef void (*esdlOperationsFactory_t)(IArrayOf<IEsdlTransformOperation> &operations, IXmlPullParser &xpp, const StringBuffer &prefix, bool withVariables, bool ignoreCodingErrors, IEsdlFunctionRegister *functionRegister);
+/**
+ * @brief Abstract interface providing external access to trace messaging in the context of an
+ *        ESDL script operation.
+ *
+ * An implementation encapsulates warning, error, and exception reporting logic for operations.
+ * Operations and their helpers can focus on why a message is being generated without having to
+ * know how messages are expected to be generated.
+ */
+interface IEsdlOperationTraceMessenger
+{
+    /**
+     * @brief Record a warning message in trace output.
+     *
+     * Implementations should not throw an exception.
+     *
+     * @param code
+     * @param msg
+     */
+    virtual void recordWarning(int code, const char* msg) const = 0;
+
+    /**
+     * @brief Record an error in trace output and possibly throw an exception.
+     *
+     * An implementation may choose to throw exceptions on any or all errors.
+     *
+     * @param code
+     * @param msg
+     */
+    virtual void recordError(int code, const char* msg) const = 0;
+
+    /**
+     * @brief Record an error in trace output and always throw an exception.
+     *
+     * An implementation is expected to throw exceptions every time.
+     *
+     * @param code
+     * @param msg
+     */
+    virtual void recordException(int code, const char* msg) const = 0;
+
+};
+
+/**
+ * @brief Encapsulation of an operation property that may be given as either an XPath or literal.
+ *
+ * Properties may be required or optional. Failure setting a required property rccords an error
+ * in a manner consistent with the containing operation's configuration.
+ *
+ * Required and optional values can be set in one of three ways:
+ *
+ * 1. Pull parser start tag and literal property name. The XPath property name is derived from the
+ *    literal name. The start tag is the source for both the XPath and the literal values.
+ * 2. Pull parser start tag, XPath property name, and literal property name. The start tag is the
+ *    source for both the XPath and the literal values.
+ * 3. An XPath property value and a literal property value.
+ *
+ * An XPath value, if not empty, takes precedence over any literal value.
+ */
+class XPathLiteralUnion
+{
+private:
+    Owned<ICompiledXpath> xpath;
+    StringAttr            literal;
+
+public:
+    inline bool isEmpty() const
+    {
+        return (!xpath && literal.isEmpty());
+    }
+
+    inline bool isXPath() const
+    {
+        return xpath;
+    }
+
+    inline bool isLiteral() const
+    {
+        return !literal.isEmpty();
```

Comment:
Actually through the usage I've seen, if xpath is set then the literal value is not used.  Therefore functionally wouldn't isLiteral() be (!isXpath() && !literal.isEmpty())?

Reply:
The set method enforces that `xpath` and `literal` cannot both be populated. Testing both is unnecessary unless there is concern of set changing to allow both values.

Reply:
Ok, yes I missed that.